src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
+        # end
+        # function (project::ProjectTo{<:$SymHerm})(dx::Tangent)
+        #     return $SymHerm(project.parent(dx.data), project.uplo)
+        # end
     end
 end
 
-# :data field
-for UL in (:UpperTriangular, :LowerTriangular)
+# Triangular
+for UL in (:UpperTriangular, :LowerTriangular, :UnitUpperTriangular, :UnitLowerTriangular)
     @eval begin
-        ProjectTo(x::T) where {T<:$UL} = ProjectTo{T}(; parent=ProjectTo(parent(x)))
-        (project::ProjectTo{<:$UL})(dx::AbstractMatrix) = $UL(project.parent(dx))
-        (project::ProjectTo{<:$UL})(dx::AbstractZero) = $UL(project.parent(dx))
-        (project::ProjectTo{<:$UL})(dx::Tangent) = $UL(project.parent(dx.data))
+        function ProjectTo(x::$UL)
+            eltype(x) == Bool && return ProjectTo(false)
+            sub = ProjectTo(parent(x))
+            return ProjectTo{$UL{eltype(eltype(sub)), eltype(sub)}}(; parent=sub)
+        end
+        (project::ProjectTo{<:$UL})(dx::AbstractArray) = $UL(project.parent(dx))
+        # (project::ProjectTo{<:$UL})(dx::AbstractZero) = $UL(project.parent(dx))
+        # (project::ProjectTo{<:$UL})(dx::Tangent) = $UL(project.parent(dx.data))
     end
 end
 
-# Transpose
-ProjectTo(x::T) where {T<:Transpose} = ProjectTo{T}(; parent=ProjectTo(parent(x)))
-function (project::ProjectTo{<:Transpose})(dx::AbstractMatrix)
-    return transpose(project.parent(transpose(dx)))
-end
-(project::ProjectTo{<:Transpose})(dx::AbstractZero) = transpose(project.parent(dx))
-
-# Adjoint
-ProjectTo(x::T) where {T<:Adjoint} = ProjectTo{T}(; parent=ProjectTo(parent(x)))
-(project::ProjectTo{<:Adjoint})(dx::AbstractMatrix) = adjoint(project.parent(adjoint(dx)))
-(project::ProjectTo{<:Adjoint})(dx::AbstractZero) = adjoint(project.parent(dx))
-
-# PermutedDimsArray
-ProjectTo(x::P) where {P<:PermutedDimsArray} = ProjectTo{P}(; parent=ProjectTo(parent(x)))
-function (project::ProjectTo{<:PermutedDimsArray{T,N,perm,iperm,AA}})(
-    dx::AbstractArray
-) where {T,N,perm,iperm,AA}
-    return PermutedDimsArray{T,N,perm,iperm,AA}(permutedims(project.parent(dx), perm))
+# Weird
+ProjectTo(x::Bidiagonal) = generic_projectto(x) # not sure!
+function (project::ProjectTo{<:Bidiagonal})(dx::AbstractMatrix)
+    uplo = LinearAlgebra.sym_uplo(project.uplo.value)
+    dv = project.dv(diag(dx))
+    ev = project.ev(uplo === :U ? diag(dx, 1) : diag(dx, -1))
+    Bidiagonal(dv, ev, uplo)
 end
-function (project::ProjectTo{P})(dx::AbstractZero) where {P<:PermutedDimsArray}
-    return P(project.parent(dx))
+
+#=
+
+x = LinearAlgebra.Tridiagonal(rand(4,4))
+backing(x) # UndefRefError: access to undefined reference
+
+=#
+
+
+# # SubArray
+# ProjectTo(x::T) where {T<:SubArray} = ProjectTo(copy(x))  # don't project on to a view, but onto matching copy
+
+# Sparse
+
+# using SparseArrays
+
+
+
+export proj_rrule
+"""
+    proj_rrule(f, args...)
+
+This calls the corresponding `rrule`, but automatically
+applies ProjectTo to all arguments.
```

Comment:
Adding this means there are two rrule-like functions:
* The one to which nothing further is done, which is the one used by AD
* The one to which projections are applied, before going to AD.

What names should these have, which should be called `rrule`? Or in other words, do we want this global application of names to apply to existing 3rd-party rule definitions (which use the name `rrule`) in which case Zygote should use `proj_rrule`. Or do we want this only to apply to rules written after it exists, in which case most of ChainRules should define methods for `raw_rrule` and this function which adds the wrapper should be `rrule`?

Next question, how does this intersect with dispatch, and more-specific rules getting precedence over less-specific? If there are two levels at which you can add a method, then it's potentially more complicated.

Reply:
I think is this is done (not convinced of this since https://github.com/JuliaDiff/ChainRulesCore.jl/pull/391/files#r665453549)
you have to define `rrule`s always.
Then Zygote etc can check if the `rrule` is defined.
And then in response can call `proj_rrule`, which might do some special projection stuff, around it's rrule, or it might do standard.

I think if we are doing this then it needs to be super rare to want to customize `proj_rrule`.
If it isn't super rare then we should just do it the way that we currently do in https://github.com/JuliaDiff/ChainRules.jl/pull/459
where the rule author decides everytime what is to be projects and what is not.

> Or do we want this only to apply to rules written after it exists, in which case most of ChainRules should define methods for raw_rrule and this function which adds the wrapper should be rrule?

Basically this one.

Reply:
This is the path least likely to surprise anyone. The method written here should then be *the* fallback method for `rrule`, and call `raw_rrule`, which has a `nothing` fallback. The two catches are:

1. won't this break Zygote's "not taking nothing seriously" story? This would be a problem with either path.

2. the dispatch is a little different, since you you define a more specialised `rrule`, it competes with this `rrule(::Any...)` not directly with the now-`raw_rrule` methods in ChainRules. Maybe that's ok.